File: app/src/main/java/com/ndmsystems/coala/CoAPResource.java
package com.ndmsystems.coala;

import com.ndmsystems.coala.message.CoAPRequestMethod;

public class CoAPResource {

    public final CoAPRequestMethod method;
    public final String path;
    private final CoAPResourceHandler handler;

    public CoAPResource(CoAPRequestMethod method, String path, CoAPResourceHandler handler) {
        this.method = method;
        this.path = path;
        this.handler = handler;
    }

    public CoAPResourceHandler getHandler() {
        return handler;
    }

    public boolean doesMatch(String path) {
        return this.path.equals(path);
    }

    public boolean doesMatch(String path, CoAPRequestMethod method) {
        return (this.path.equals(path) && this.method.equals(method));
    }

    @Override
    public boolean equals(Object other) {
        if (other == null) return false;
        if (other == this) return true;
        if (!(other instanceof CoAPResource)) return false;
        CoAPResource otherResource = (CoAPResource) other;
        return method == otherResource.method && path.equals(otherResource.path);
    }

    abstract public static class CoAPResourceHandler {
        abstract public CoAPResourceOutput onReceive(CoAPResourceInput inputData);
    }
}

File: app/src/main/java/com/ndmsystems/coala/CoAPResource.kt
package com.ndmsystems.coala

import com.ndmsystems.coala.message.CoAPRequestMethod

open class CoAPResource(val method: CoAPRequestMethod, val path: String, val handler: CoAPResourceHandler) {

    fun doesMatch(path: String): Boolean {
        return this.path == path
    }

    fun doesMatch(path: String, method: CoAPRequestMethod): Boolean {
        return this.path == path && this.method == method
    }

    override fun equals(other: Any?): Boolean {
        if (other == null) return false
        if (other === this) return true
        if (other !is CoAPResource) return false
        return method === other.method && path == other.path
    }

    abstract class CoAPResourceHandler {
        abstract fun onReceive(inputData: CoAPResourceInput): CoAPResourceOutput
    }
}